_source/_posts/2021-10-16-full-stack-java.md

```diff
@@ -0,0 +1,693 @@
+
+If you upload an image you took with your smartphone, the height, width, and taken values should all be populated. If they're not, chances are your image doesn't have the data in it.
+
+> Need some sample photos with EXIF data? You can download pictures of my 1966 VW Bus from [an album on Flickr](https://www.flickr.com/photos/mraible/albums/72157689027458320).
+
+## Add a React Photo Gallery
+
+You've added metadata extraction to your backend, but your photos still display in a list rather than in a grid (like Flickr). To fix that, you can use the [React Photo Gallery](https://github.com/neptunian/react-photo-gallery) component. Install it using npm:
+
+```bash
+npm i react-photo-gallery@8
+```
+
+In `src/main/webapp/app/entities/photo/photo.tsx`, add an import for `Gallery`:
+
+```ts
+import Gallery from 'react-photo-gallery';
+```
+
+Then add a `photoSet` variable in the `render()` method, and the `<Gallery>` component right after the closing `</h2>`.
+
+```ts
+render() {
+  const { photoList, match } = this.props;
+  const photoSet = photoList.map(photo => ({
+    src: `data:${photo.imageContentType};base64,${photo.image}`,
+    width: photo.height > photo.width ? 3 : photo.height === photo.width ? 1 : 4,
+    height: photo.height > photo.width ? 4 : photo.height === photo.width ? 1 : 3
+  }));
+
+  return (
+    <div>
+      <h2 id="photo-heading">
+        ...
+      </h2>
+      <Gallery photos={photoSet} />
+      ...
+  );
+}
+```
+
+Since you only modified the front-end code, you can run `npm start` to start an instance of webpack-dev-server that proxies requests to the backend and auto-refreshes your browser (using [Browsersync](https://browsersync.io/)) every time you change any React files.
+
+Log in and navigate to **Entities** > **Photos** in the top nav bar. You should be able to upload photos and see the results in a nice grid at the top of the list.
+
+//todo: update screenshot during QA
+
+{% img blog/full-stack-java/photo-gallery.png alt:"Gallery with Photos" width:"800" %}{: .center-image }
+
+//todo: update GitHub links
+You can also add a "lightbox" feature to the grid so you can click on photos and zoom in. The [React Photo Gallery docs](https://neptunian.github.io/react-photo-gallery/) show how to do this. I've integrated it into the example for this post, but I won't show the code here for the sake of brevity. You can see the [final `photo.tsx` with Lightbox added on GitHub](https://github.com/oktadeveloper/okta-react-photo-gallery-example/blob/master/src/main/webapp/app/entities/photo/photo.tsx) or a [diff of the changes necessary](https://github.com/oktadeveloper/okta-react-photo-gallery-example/commit/47f9ceab2b00f1d7f41d286686c9159f79decc11).
+
+## Make Your Full Stack Java App Into a PWA
```

Comment:
I'm thinking whether you should expand the PWA shortcut in the heading. If I'm not wrong it wasn't used before in this article, so some readers may have problems with that, what do you think?